Added hibernate.

OleksandrOrlovIT opened the pull request:
![build2](https://github.com/user-attachments/assets/f0ea9f81-73d9-4aad-aa61-671aae562e63)

![Jacoco2](https://github.com/user-attachments/assets/20743ab7-840f-4593-9f58-5f4ee2517d4b)

OleksandrOrlovIT committed:
Added hibernate to the project. Added env library to load env from .env file. Closes #14

OleksandrOrlovIT committed:
Changed entities to match technical requirements. Deleted storage file. Closes #15 .

OleksandrOrlovIT committed:
Created dao layer with hibernate. Closes #16

OleksandrOrlovIT committed:
Created Trainer Service. Closes #17

OleksandrOrlovIT committed:
Created TraineeService. Closes #19

OleksandrOrlovIT committed:
Completed last requirements. Closes #20

Anderwerty commented on src/main/java/orlov/programming/springcoregym/bootstrap/TrainingTypeLoader.java:
```diff
@@ -0,0 +1,26 @@
+
+    private final TrainingTypeService trainingTypeService;
+
+    public void loadDefaultTrainingTypes() {
```
why not use just sql script to execute?

Anderwerty commented on src/main/java/orlov/programming/springcoregym/configuration/AppConfig.java:
```diff
@@ -1,17 +1,88 @@
 @ComponentScan(basePackages = "orlov.programming.springcoregym")
 public class AppConfig {
+
+    private final Dotenv dotenv;
```
why do we need it?

Anderwerty commented on src/main/java/orlov/programming/springcoregym/configuration/AppConfig.java:
```diff
@@ -1,17 +1,88 @@
+import java.util.Objects;
+import java.util.Properties;
 
 @Log4j2
```
do we need this annotation here?

Anderwerty commented on src/main/java/orlov/programming/springcoregym/configuration/EnvConfig.java:
```diff
@@ -0,0 +1,22 @@
+ * Configuration class for loading environment variables.
+ */
+@Configuration
+public class EnvConfig {
```
Why not use spring functionality instead of this third-party class?

Anderwerty commented on src/main/java/orlov/programming/springcoregym/dao/AbstractDao.java:
```diff
@@ -0,0 +1,51 @@
+
+import java.util.List;
+
+public abstract class AbstractDao<T, ID> implements Dao<T, ID>{
```
formatting issue: should be a space before brace

Anderwerty commented on src/main/java/orlov/programming/springcoregym/dao/AbstractDao.java:
```diff
@@ -0,0 +1,51 @@
+    }
+
+    @Override
+    public T findById(ID id) {
```
let's make such method return Optional

Anderwerty commented on src/main/java/orlov/programming/springcoregym/dao/AbstractDao.java:
```diff
@@ -0,0 +1,51 @@
+        return entityManager.find(getEntityClass(), id);
+    }
+
+    public List<T> findAll() {
```
should here be override annotation?

Anderwerty commented on src/main/java/orlov/programming/springcoregym/dao/impl/user/trainee/TraineeDao.java:
```diff
@@ -1,7 +1,14 @@
+import java.util.List;
+
+public interface TraineeDao extends DaoUsernameFindable<Trainee, Long> {
+    List<Training> getTrainingsByDateUsernameTrainingType
```
if a method has more than 3 parameters better to encapsulate them into some dto

Anderwerty commented on src/main/java/orlov/programming/springcoregym/dao/impl/user/trainee/TraineeDaoImpl.java:
```diff
@@ -1,49 +1,71 @@
+            return Optional.of(trainee);
+
+        } catch (NoResultException e) {
+            log.info("No trainee found with username: {}", username);
```
why do we need this logger here?

Anderwerty commented on src/main/java/orlov/programming/springcoregym/dao/impl/user/trainee/TraineeDaoImpl.java:
```diff
@@ -1,49 +1,71 @@
+        query.setParameter("username", username);
+        int rowsAffected = query.executeUpdate();
+
+        log.info("Deleted {} trainee(s) with username: {}", rowsAffected, username);
```
do we need it?
maybe it is better to make such a method to return a boolean value (true - deleted, false - was not deleted)

Anderwerty commented on src/main/java/orlov/programming/springcoregym/facade/GYMFacade.java:
```diff
@@ -4,16 +4,31 @@
+import java.time.LocalDate;
+import java.util.List;
+
 public interface GYMFacade {
```
It looks like a violation of  SOLID (S-single responsibility)

Anderwerty commented on src/main/java/orlov/programming/springcoregym/facade/impl/GYMFacadeImpl.java:
```diff
@@ -1,70 +1,248 @@
+            return traineeService.create(trainee);
+        } catch (Exception e) {
+            log.error(e);
+            return null;
```
It is bad practice to return null
https://medium.com/javarevisited/just-dont-return-null-dcdf5d77128f

Anderwerty commented on src/main/java/orlov/programming/springcoregym/model/training/Training.java:
```diff
@@ -1,22 +1,76 @@
+    private Long trainingDuration;
+
+    @Override
+    public String toString() {
```
why not use Lombok to generate toString, equals, hashCode methods?

OleksandrOrlovIT replied:
Using Lombok with hibernate entity relationships results in many pitfalls and inconsistencies. If it's not a cornerstone, I would rather stick to the JPABuddy implementation than the lombok one.

Anderwerty commented on src/main/java/orlov/programming/springcoregym/service/CRUDService.java:
```diff
@@ -1,5 +1,5 @@
 
-public interface CRUDService<T> extends UpdatableService<T>{
-    void delete(T t);
+public interface CRUDService<E, ID> extends UpdatableService<E, ID>{
```
Logic in services is not so straightforward, so no need to use such Interfaces as UpdableService,...
Better to create separate contracts for each service.

Anderwerty commented on src/main/java/orlov/programming/springcoregym/service/user/trainer/TrainerServiceImpl.java:
```diff
@@ -83,4 +94,101 @@ private void checkFirstLastNames(Trainer trainer){
+            throw new IllegalArgumentException("Trainer not found " + username);
+        }
+
+        return foundTrainer.get().getPassword().equals(password);
```
can we get NPE here?

Anderwerty commented on src/test/java/orlov/programming/springcoregym/bootstrap/TrainingTypeLoaderTest.java:
```diff
@@ -0,0 +1,31 @@
+
+        assertEquals(5, trainingTypeService.findAll().size());
+    }
+}
```
please add a blank line after last brace

OleksandrOrlovIT committed:
Changed Java code to sql code in order to populate db. Changed thirdParty class to spring properties. Changed all possible null methods to return optionals. Created dto to fix method with 4 params. Created several Facades instead of one bif facade for single responsibility principle. Deleted wrong service tree interfaces.

OleksandrOrlovIT committed:
Added constant for strings methods in dao layer.
@transactional deleteByUsername
Deleted log.info findByUsername dao
findByids IllegalArgumentException
added newline to new facade classes
fixed entities using Lombok
threadLocal instead of simple User class in authenticationService
changed getTrainingByDateTraineeName
decided to use get as a main verb for methods
all methods start from verb
10 to const in services
in optional change ifs to orElseThrow. #23

OleksandrOrlovIT committed:
Pageable addd to getTrainersWithoutPassedTrainee
properties deleted update for hibernateProperties
changed config to read property files
changed test names to use camelCase and start with method name
added assertTrue when create entity in tests. Closes #23

OleksandrOrlovIT committed:
Added javadoc to the dao and service interfaces. Added a bit of tests. Closes #22
